### DISCUSSION TITLE
Changing Inertia Matrix and Mass at runtime

### DISCUSSION THREAD

**Discussion by Dobid** [JSBSim-Team/jsbsim#1029] [Q&A]
Hello JSBSimTeam,

In the same spirit as the question I previously asked on #990 I would like to be able to change the inertia matrix and mass parameters by changing a property.

Following the answer given in #990 I attempted this:
```xml
    <mass_balance>
        <property value="1.2290">fdm/inertia-matrix/ixx</property>
        <property value="0.1702">fdm/inertia-matrix/iyy</property>
        <property value="0.8808">fdm/inertia-matrix/izz</property>
        <property value="0.9343">fdm/inertia-matrix/ixz</property>
        <property value="3.3640">fdm/emptywt</property>

        <ixx unit="KG*M2"><property>fdm/inertia-matrix/ixx</property></ixx>
        <iyy unit="KG*M2"><property>fdm/inertia-matrix/iyy</property></iyy>
        <izz unit="KG*M2"><property>fdm/inertia-matrix/izz</property></izz>
        <ixz unit="KG*M2"><property>fdm/inertia-matrix/ixz</property></ixz>
        <emptywt unit="KG"><property>fdm/emptywt</property></emptywt>
        <location name ="CG" unit="M">
            <x> 0.0 </x>
            <y> 0.0 </y>
            <z> 0.0 </z>
        </location>
    </mass_balance>
```
And I got the following error:
```
In file fdm_descriptions/aircraft/x8/x8.xml: line 58
Expected numeric value, but got no data
terminate called after throwing an instance of 'std::length_error'
  what():
In file fdm_descriptions/aircraft/x8/x8.xml: line 58
Expected numeric value, but got no data
```
Is there a way to do so ?

Thanks !

**Comment by seanmcleod** [comment]
Yep, currently the inertia values can't be specified as properties. Only way at the moment to adjust them at run-time is to add/remove point masses.

**Reply by Dobid** [reply]
Alright thanks !

**Comment by bcoconni** [comment]
Sorry to join the conversation a bit late, but I would like to add some points:
* For the record, this topic has already been discussed in the discussion #961 in which it has been highlighted that adding/moving/removing point masses do not allow to represent all possible inertia matrices (thanks @scotteh432 for having done the maths). Definitely worth of read.
* In addition, allowing the inertia matrix to be defined by properties would certainly need some thoughts. Especially since the equation of motion (EOM) for attitude dynamics is
$$\frac{dL}{dt}=M$$
where $L=J\omega$ is the [angular momentum](https://en.wikipedia.org/wiki/Angular_momentum) and $M$ the moment applied to the body. In JSBSim, the inertia matrix $J$ is assumed to be constant and the EOM above simplifies into
$$J\frac{\omega}{dt}=M$$
However if the inertia matrix is allowed to change over time, we'd need to account for its derivative and the exact EOM would then be
$$\frac{dJ}{dt}\omega+J\frac{\omega}{dt}=M$$
Of course if the rotation rate $\omega$ and/or $\frac{dJ}{dt}$ are low, we can safely ignore the term $\frac{dJ}{dt}\omega$. And I'd guess that this assumption stands for all the models that use the point masses trick. However adding the support of a mutable inertia matrix to JSBSim would most certainly imply that we manage the term $\frac{dJ}{dt}\omega$ and that would be another story.

**Comment by seanmcleod** [comment]
> And I'd guess that this assumption stands for all the models that use the point masses trick.

Plus for fuel burn, since this also modifies the inertia matrix.

**Reply by bcoconni** [reply]
> Plus for fuel burn, since this also modifies the inertia matrix.

True. The fuel flow of aircraft engines is low (some kg/s at worse for airliners) and is negligible with respect to the aircraft total mass and inertia (tens or hundreds of metric tons for an airliner). So we can safely ignore the $\frac{dJ}{dt}\omega$ term.

Also - for the record - the equivalent term $\frac{dm}{dt}v$ in the 2nd law of Newton is accounted for in the thrust (from jet engines where the ejected mass of fuel is a few percent of the ejected mass flow, to rocket engines where the ejected mass of fuel is 100% of the ejected mass).

**Comment by seanmcleod** [comment]
Yep, I guess something like a SpaceX rocket would have a large $\frac{dJ}{dt}$ but typically would have a small $\omega$ unless it was spinning out of control :wink:

**Reply by bcoconni** [reply]
Good point !
